### DISCUSSION TITLE
why nodeConfVolume is needed in RedisCluster?

### DISCUSSION THREAD

**Discussion by wkd-woo** [OT-CONTAINER-KIT/redis-operator#871] [General]
**Describe your question**
Looking at the 'RedisCluster' configuration, there is an option for nodeConfVolume.

I wonder why you classified using nodeConfVolume separately without using the default volume(reclaimed by volumeClaimTemplate). I've looked up a lot of repository, but I couldn't find any mention of this, so I'm inquiring.

**What version of redis-operator are you using?**
redis-operator version: v0.15.0

**Comment by solipsistadventurer** [comment]
I was jus looking at this today :-). I believe it is to facilitate persistence of the nodes.conf file in case a pod is replaced. ([documentation](https://redis.io/docs/management/scaling/#create-a-redis-cluster)) It keeps record of node UIDs in the cluster, including the "local' node. By using persistant storage for this file, a node can be replaced without the need to reconfigure the cluster.

Note that I am not very familiar with Redis, nor did I test my theory extensively.

**Comment by wkd-woo** [comment]
I think you're right, too.

However, when you set nodeConfVolume (operator v0.15.0 with redis v7.0.12) PVC of the default volume is not created.

I just checked the redis image of OT-CONTAINER-KIT, and I think it's because only one volume is mounted (/data) in the container,
Is this intended? In the image of redis v7.2.3, two are mounted. (/data, /node-conf)


```
ENTRYPOINT ["/usr/bin/entrypoint.sh"]
USER 1000
EXPOSE map[6379/tcp:{}]
WORKDIR /data
VOLUME [/data]
RUN /bin/sh -c chown -R 1000:0 /etc/redis && chmod -R g+rw /etc/redis && mkdir /data && chown -R 1000:0 /data && chmod -R g+rw /data # buildkit

...

```
Above is image layers in [quay.io/opstree/redis:v7.0.12](https://quay.io/repository/opstree/redis/manifest/sha256:22061f566331bbe3dc7c42cfe79fbf5e1ed318e5ee63648ec6ebd79f0ebe4fbe)


```
ENTRYPOINT ["/usr/bin/entrypoint.sh"]
USER 1000
EXPOSE map[6379/tcp:{}]
WORKDIR /data
VOLUME [/node-conf]
VOLUME [/data]
RUN |1 TARGETARCH=amd64 /bin/sh -c chown -R 1000:0 /etc/redis && chmod -R g+rw /etc/redis && mkdir /data && chown -R 1000:0 /data && chmod -R g+rw /data && mkdir /node-conf && chown -R 1000:0 /node-conf && chmod -R g+rw /node-conf && chmod -R g+rw /var/run # buildkit

...

```
Above is image layers in [quay.io/opstree/redis:v7.2.3](https://quay.io/repository/opstree/redis/manifest/sha256:2c7380cabdd1b7a999c58370d81808cd2375bba2df9e1d330ad0e757ef4528fb)
There are two volumes. (/data, /node-conf)

<br>

---

So, if we set `nodeConfVolume: true`, only one PVC is actually bound (in redis v7.0.7).

```
$ kubectl get statefulset.apps/redis-cluster-leader -o yaml
apiVersion: apps/v1
kind: StatefulSet
...

  volumeClaimTemplates:
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      annotations:
        meta.helm.sh/release-name: redis-cluster
        meta.helm.sh/release-namespace: redis-operator
        redis.opstreelabs.in: "true"
        redis.opstreelabs.instance: redis-cluster
      creationTimestamp: null
      labels:
        app: redis-cluster-leader
        app.kubernetes.io/component: middleware
        app.kubernetes.io/instance: redis-cluster
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: redis-cluster
        app.kubernetes.io/version: 0.15.1
        dbservice: redis
        helm.sh/chart: redis-cluster-0.15.11
        redis_setup_type: cluster
        region: kr2
        role: leader
        topology: cluster
      name: node-conf
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 10Gi
      selector:
        matchLabels:
          dbservice: redis
          topology: cluster
      storageClassName: local-storage
      volumeMode: Filesystem
    status:
      phase: Pending

...

$ kubectl get pvc
NAME                                 STATUS   VOLUME               CAPACITY   ACCESS MODES   STORAGECLASS    AGE
node-conf-redis-cluster-follower-0   Bound    conf-follower-pv-0   10Gi       RWO            local-storage   16h
node-conf-redis-cluster-follower-1   Bound    conf-follower-pv-1   10Gi       RWO            local-storage   16h
node-conf-redis-cluster-follower-2   Bound    conf-follower-pv-2   10Gi       RWO            local-storage   16h
node-conf-redis-cluster-leader-0     Bound    conf-leader-pv-0     10Gi       RWO            local-storage   16h
node-conf-redis-cluster-leader-1     Bound    conf-leader-pv-1     10Gi       RWO            local-storage   16h
node-conf-redis-cluster-leader-2     Bound    conf-leader-pv-2     10Gi       RWO            local-storage   16h
```

= That means we can't use the default volume at the same time.



In redis v7.2.3, nodeConfVolume and defaultVolume are available at the same time.

**Comment by drivebyer** [comment]
In my understanding, Redis can function as a cache system. In this scenario, Redis does not require PVC but it does need a persistent node.conf alone.
